### ISSUE TITLE
[Feature] Add transforms for randomly converting image to grayscale

### ISSUE THREAD

**Issue by sourabhd** [pytorch/vision#299]
Data augmentation (transform) for randomly converting image into grayscale (with probability p) is useful for handling datasets containing a mix of rgb and grayscale images. 

Please check my implementation [here](https://github.com/sourabhd/vision/commit/a93e53a850ee923dd090ee5a960fadd1a8bd8c76?diff=split)
Let me know if this looks reasonable addition. I could send a pull request for the same.

**Comment by sourabhd** [comment]
@soumith @apaszke @fmassa

**Comment by alykhantejani** [comment]
Hi @sourabhd, 

I've taken a quick look at the code and it seems like you convert an image to grayscale and then back to RGB (repeating the grayscale image 3 times).

I'm not sure when you would want to do this, i.e. if your dataset is a mix of RGB and grayscale images I would think you would either want all grayscale images (single channel) or to convert the grayscale ones to 3-channel grayscale and mix these with the RGB ones. 

A mix of single channel and 3 channel images wouldn't make sense as your network needs to know the number of input channels?

**Comment by sourabhd** [comment]
@alykhantejani 

1. A 3-channel image could be grayscale if R == G == B
[Example image from ImageNet](http://farm1.static.flickr.com/132/387212290_4573a5f996.jpg)
[Grayscale in RGB](https://www.w3schools.com/colors/colors_shades.asp
)
2. Origins
- Dataset might contain such 3-channel images like the example shown above

- If dataset has single channel images, they need to be converted to 3 channel as pre-trained models are available only for 3 channel. This is usually done by replicating across channels. This leads to the same situation

3. Why is it needed ?
Labeling of dataset could be expensive (example FACS, emotions etc) and we want to make use of both the colored and grayscale images (instead of throwing away one set).  For introducing, invariance to grayscaling of an image, we could employ an augmentation where we randomly (with a probability p) grayscale an image. The idea is that over multiple epochs, the network sees an image as well as its grayscale counterpart (which has the same label) and learns the invariance during backprop accordingly.

**Comment by alykhantejani** [comment]
So I think that this could go both ways i.e. the user wants to change 3-channel images to single channel grayscale or the user wants to change 1-channel images to 3-channel ones. I think a `to_grayscale` function can be misleading in this case as sometimes is returns 1-channel images and sometimes 3-channel. 

Additionally, depending on the users preference, it should be easy for them to encapsulate their desired behavior into a function (using PIL's convert/stack) and chain these together in a `Compose`? 

@fmassa wdyt?

**Comment by sourabhd** [comment]
@alykhantejani In that case we could have two functions to_grayscale_singlechannel and to_grayscale_threechannel.

**Comment by alykhantejani** [comment]
@sourabhd yeah, I'd be happy with a `to_grayscale` function/transform with a `num_output_channels` kwarg. Can you send a PR?

**Comment by alykhantejani** [comment]
Fixed via #325
